low-latency-networktables.adoc

```diff
@@ -0,0 +1,181 @@
+the vast majority of values (such as configuration parameters) this
+latency is not an issue.  However, it is often desirable to update a
+handful of values in a more real-time, low latency fashion, such as
+when doing image processing (where a new frame may be processed every
+30 ms).
+
+== Design
+
+=== Protocol Specification Changes
+
+This section describes the required changes to the
+{networktables3}[NetworkTables 3.0 Protocol].  There are no changes to
+the initial handshake (other than the version number) or other TCP
+operations.  The only operational change is to send Entry Update
+messages for Low-Latency flagged entries via UDP rather than TCP.
+
+==== Protocol Revision
+
+Protocol revision greater than or equal to 3.1 (0x0301) indicate
+support for low-latency messaging.
+
+==== Transport Layer
+
+The Server should listen on UDP port 1735 for incoming low-latency
+messages.  The Client should listen on a UDP port starting at port
+1735 (to allow for multiple Clients on a single machine, higher port
+numbers, e.g. 1736, 1737, etc. may need to be used).
+
+==== Client Hello
+
+Add client UDP port to Client Hello message (when connecting with
+protocol revision 3.1).
+
+[cols="1,3"]
+|===
+|Field Name |Field Type
+
+|0x01 - Client Hello
+|1 byte, unsigned; Message Type
+
+|Client Protocol Revision
+|2 bytes, Unsigned 16-bit integer (big-endian). See
+<<def-protocol-revision,Protocol Revision>>.
+
+|Client UDP Port
+|2 bytes, Unsigned 16-bit integer (big-endian).
+
+|Client identity (name)
+|<<entry-value-string,String>>
+|===
+
+==== Client Updates an Entry
+
+When User Code on a Client updates the value of an Entry with the Low
+Latency flag set, the Client should send the associated Entry Update
+mesage to the Server via UDP, rather than TCP.  The Client should only
+send Entry Update messages for these Entries via TCP under the
+following conditions:
+
+* The Server is not version 3.1 or greater.
+
+* The Entry Update message would not fit within the MTU of the
+transmitting medium.  MTU path discovery is not required.  The Client
+should assume the MTU is no greater than 1500 bytes.
+
+When generating UDP packets, the Client should pack as many Entry
+Updates as possible within a single packet (respecting MTU), but the
+Entry Update messages must be aligned to each generated packet (e.g.
+the packet must start with the start of an Entry Update message).
+
+==== Server Updates an Entry
+
+When User Code on the Server updates the value of an Entry with the
+Low Latency flag set, the Server should send the associated Entry
+Update message via UDP, rather than TCP, to version 3.1 or greater
+Clients, in accordance with the previous section.
+
+==== Entry Flags
+
+Entry Flags are updated to add a low-latency flag.  The flags are
+defined such that an Entry cannot be both Low Latency and Persistent.
+The flags definition is backwards compatible with version 3.0's
+persistent flag.
+
```

Comment:
Why not allowed to be Persistent and Low Latency? It looks like below the bit mask is being set up correctly to allow both.